third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor_v2.rs

```diff
@@ -0,0 +1,1883 @@
+    fn move_ref(&mut self, dest: TempIndex, src: TempIndex) {
+        let Some(label) = self.local_to_label_map.remove(&src) else {
+            return;
+        };
+        self.local_to_label_map.insert(dest, label);
+        self.check_graph_consistency()
+    }
+
+    /// Copies a reference from source to destination. This create a new lifetime node and clones the edges
+    /// leading into the node associated with the source reference.
+    fn copy_ref(
+        &mut self,
+        _loc: Loc,
+        dest: TempIndex,
+        src: TempIndex,
+        _code_offset: CodeOffset,
+        _qualifier: u8,
+    ) {
+        if let Some(label) = self.label_for_local(src) {
+            self.local_to_label_map.insert(dest, *label);
+        }
+        /*
+        let new_label = self.make_local(dest, code_offset, qualifier, false);
+        let Some(curr_label) = self.label_for_local(src).cloned() else {
+            return;
+        };
+        let mut new_edges = vec![];
+        for (parent, edge) in self.parent_edges(&curr_label) {
+            new_edges.push((parent, BorrowEdge {
+                kind: edge.kind.clone(),
+                is_mut: edge.is_mut,
+                loc: edge.loc.clone(), // use the location of original borrow
+                target: new_label,
+            }));
+        }
+        for (parent, new_edge) in new_edges {
+            self.add_edge(parent, new_edge)
+        }
+        self.check_graph_consistency()
+         */
+    }
+
+    /// Returns an iterator of the edges which are leading into this node.
+    #[allow(unused)]
+    fn parent_edges<'a>(
+        &'a self,
+        label: &'a LifetimeLabel,
+    ) -> impl Iterator<Item = (LifetimeLabel, &'a BorrowEdge)> + '_ {
+        self.node(label).parents.iter().flat_map(move |parent| {
+            self.children(parent)
+                .filter(move |edge| &edge.target == label)
+                .map(|e| (*parent, e))
+        })
+    }
+
+    /// Returns the roots of this node, that is those transitive parents which have no further parents.
+    fn roots(&self, label: &LifetimeLabel) -> BTreeSet<LifetimeLabel> {
+        let mut roots = BTreeSet::new();
+        let mut todo = self.node(label).parents.iter().cloned().collect::<Vec<_>>();
+        while let Some(l) = todo.pop() {
+            let mut parents = self.node(&l).parents.iter().cloned().collect::<Vec<_>>();
+            if parents.is_empty() {
+                // Found a root
+                roots.insert(l);
+            } else {
+                // Explore parents
+                todo.append(&mut parents)
+            }
+        }
+        roots
+    }
+
+    /// Returns the locals borrowed
+    pub fn borrowed_locals(&self) -> impl Iterator<Item = TempIndex> + '_ {
+        self.local_to_label_map.keys().cloned()
+    }
+
+    /// Checks if the given local is borrowed
+    pub fn is_borrowed(&self, temp: TempIndex) -> bool {
+        self.borrowed_locals().contains(&temp)
```

Comment:
Does this require (potentially) all elements in the iterator returned from `borrowed_locals()` to be cloned to check for element containment?

Reply:
This has been changed in the next PR already